design/backup-resource-list.md

```diff
@@ -0,0 +1,92 @@
+
+After performing a backup, a new file will be created that contains the list of the resources that have been included in the backup.
+This file will be persisted in object storage alongside the backup contents and existing metadata.
+
+A section will be added to the output of `velero backup describe <name> --details` command to view this metadata.
+
+## Detailed Design
+
+### Metadata file
+
+This metadata will be in JSON (or YAML) format so that it can be easily inspected from the bucket outside of Velero tooling, and will contain the API resource and group, namespaces and names of the resources:
+
+```
+deployments.apps:
+- default/database
+- default/wordpress
+services:
+- default/database
+- default/wordpress
+secrets:
+- default/database-root-password
+- default/database-user-password
+configmaps:
+- default/database
+```
+
+The filename for this metadata will be `<backup name>-resource-list.json`.
+The top-level key is the string form of the `schema.GroupResource` type that we currently keep track of in the backup controller code path.
+
+### Changes in Backup controller
+
+The Backupper currently initialises a map to track the `backedUpItems` (https://github.com/heptio/velero/blob/1594bdc8d0132f548e18ffcc1db8c4cd2b042726/pkg/backup/backup.go#L269), this is passed down through GroupBackupper, ResourceBackupper and ItemBackupper. Moving the map initialisation to the BackupController will provide access to the resulting list there after a successful backup.
+
+The `backedUpItems` map is currently a flat structure and will need to be converted to the nested structure above, grouped by `schema.GroupResource`.
+
+After converting to the right format, it can be passed to the `persistBackup` function to persist the file in object storage.
+
+### Changes to DownloadRequest CRD and processing
+
+A new `DownloadTargetKind` "BackupResourceList" will be added to the DownloadRequest CR.
+
+The `GetDownloadURL` function in the `persistence` package will be updated to handle this new DownloadTargetKind to enable the Velero client to fetch the metadata from the bucket.
+
+### Changes to `velero backup describe <name> --details`
+
+This command will need to be updated to fetch the metadata from the bucket using the `Stream` method used in other commands.
+The file will be read in memory and displayed in the output of the command.
+Depending on the format the metadata is stored in, it may need processing to print in a more human-readable format.
+If we choose to store the metadata in YAML, it can likely be directly printed out.
+
+If the metadata file does not exist, this is an older backup and we cannot display the list of resources that were backed up.
+
+## Open Questions
+
+- Do we want to show the group version in the metadata (i.e. do we want users to see `apps/v1/Deployment` instead of `deployments.apps`)? If so, we wouldn't be able to use the existing list of backedUpItems list, or we'd need to change it to track schema.GroupVersionKinds instead.
```

Comment:
I think this would be nice to have -- particularly for the use case where a user is restoring an older backup into a new cluster, and needs to ensure that the backed-up API versions exist in their new cluster before restoring. What do others think?

Reply:
That's a very good point, do you think we can change backedUpItems to track with versions? I wasn't sure if it relies on unambiguous versions to make sure there are no duplicates?

Reply:
Velero only backs up the server's "preferred version" of each API group, so in theory we should only have one version per resource. We could update the `itemKey` struct to use a GVR rather than just a GR, which should work OK (need to figure out details though).

Reply:
Agreed that writing the version of the backed up item would be nice, especially since we should only have one per item.